File: leetcode-scala/src/dfs/easy/BalancedBinaryTree.scala
package dfs.easy

import tools.TreeNode

object BalancedBinaryTree {

    /**
     * 620 ms   72.22%
     * 51.6 MB  100.00%
     * @param root
     * @return
     */
    def isBalanced(root: TreeNode): Boolean = {
        if (root == null) return true
        if (Math.abs(maxDepth(root.left) - maxDepth(root.right)) > 1) return false
        isBalanced(root.left) && isBalanced(root.right)
    }

    def maxDepth(root: TreeNode): Int = {
        if (root == null) return 0
        if (root.left == null && root.right == null) return 1
        Math.max(maxDepth(root.left), maxDepth(root.right)) + 1
    }

}

File: leetcode-java/src/classify/recursion/BalancedBinaryTree.java
package classify.recursion;

import tools.TreeNode;

/**
 * @author yutiantang
 * @create 2021/3/31 22:13
 */
public class BalancedBinaryTree {

    /**
     * 1ms      99.97%
     * 38.7MB   22.72%
     * @param root
     * @return
     */
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }

        if (Math.abs(maxDepth(root.left) - maxDepth(root.right)) > 1) {
            return false;
        }

        return isBalanced(root.left) && isBalanced(root.right);
    }

    private int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }
        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }


    /**
     * 官方解题
     * @param root
     * @return
     */
    public boolean isBalanced2(TreeNode root) {
        return height(root) >= 0;
    }

    public int height(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftHeight = height(root.left);
        int rightHeight = height(root.right);
        if (leftHeight == -1 || rightHeight == -1 || Math.abs(leftHeight - rightHeight) > 1) {
            return -1;
        } else {
            return Math.max(leftHeight, rightHeight) + 1;
        }
    }
}
